### ISSUE TITLE
Mapping between dictionaries and lists

### ISSUE THREAD

**Issue by blemasle** [agileobjects/AgileMapper#173]
Hi (again),

I'm trying to convert a dictionary mapped over a list. Ideally, I would like to set the key to one of the destination property, but I can get around that with the [workaround describe here](https://github.com/agileobjects/AgileMapper/issues/162#issuecomment-541199626).

Mapping between dictionaries work, but mapping between a dictionary and a list yields an empty list in the best case scenario.  In the worst case, an `NullReferenceException` is raised

> System.NullReferenceException: 'Object reference not set to an instance of an object.'
> typePair was null.
>at AgileObjects.AgileMapper.Members.MemberMapperDataExtensions.TypesMatch(ITypePair typePair, IList`1 contextTypes) in C:\Users\Bertrand\source\extern\AgileMapper\AgileMapper\Members\MemberMapperDataExtensions.cs:line 496
>at AgileObjects.AgileMapper.Members.MemberMapperDataExtensions.GetAppropriateMappingContextAccess(IMemberMapperData mapperData, Type[] contextTypes) in C:\Users\Bertrand\source\extern\AgileMapper\AgileMapper\Members\MemberMapperDataExtensions.cs:line 447
>at AgileObjects.AgileMapper.Configuration.ParametersSwapper.SwapArgs.GetAppropriateMappingContextAccess(Type[] contextTypes) in C:\Users\Bertrand\source\extern\AgileMapper\AgileMapper\Configuration\ParametersSwapper.cs:line 391
>at AgileObjects.AgileMapper.Configuration.ParametersSwapper.SwapArgs.GetAppropriateMappingContextAccess() in C:\Users\Bertrand\source\extern\AgileMapper\AgileMapper\Configuration\ParametersSwapper.cs:line 388
>at AgileObjects.AgileMapper.Configuration.ParametersSwapper.GetAppropriateMappingContext(SwapArgs swapArgs) in C:\Users\Bertrand\source\extern\AgileMapper\AgileMapper\Configuration\ParametersSwapper.cs:line 183
>at AgileObjects.AgileMapper.Configuration.ParametersSwapper.ReplaceParameters(SwapArgs swapArgs, Func`2[] parameterFactories) in C:\Users\Bertrand\source\extern\AgileMapper\AgileMapper\Configuration\ParametersSwapper.cs:line 166
>at AgileObjects.AgileMapper.Configuration.ParametersSwapper.SwapForSourceAndTarget(SwapArgs swapArgs) in C:\Users\Bertrand\source\extern\AgileMapper\AgileMapper\Configuration\ParametersSwapper.cs:line 151
>at AgileObjects.AgileMapper.Configuration.ParametersSwapper.Swap(LambdaExpression lambda, Type[] contextTypes, IMemberMapperData mapperData, Func`4 targetValueFactory) in C:\Users\Bertrand\source\extern\AgileMapper\AgileMapper\Configuration\ParametersSwapper.cs:line 286

See [this fiddle](https://dotnetfiddle.net/FzZgJC).

```csharp
using System;
using AgileObjects.AgileMapper;
using System.Collections.Generic;
		
public class Program
{
    public static void Main()
	{
		Mapper
				.WhenMapping
				.From<A>()
				.To<C>()
				.Map((src, dst) => src.Something).To(dst => dst.Value);

			//raise an NullReferenceException if enabled
			//Mapper
			//	.WhenMapping
			//	.FromDictionariesWithValueType<A>()
			//	.To<List<C>>()
			//	.Map((src, dst) => src.Values.ToList()).ToTarget();

			Dictionary<string, A> a = new Dictionary<string, A>
			{
				{ "test", new A { Something = "value" } },
				{ "test2", new A { Something = "value2" } },
			};

			Dictionary<string, C> c = Mapper
				.Map(a).ToANew<Dictionary<string, C>>();
			Console.WriteLine("Between dictionaries : " + c.Count);

			List<C> c2 = Mapper
				.Map(a).ToANew<List<C>>();
			Console.WriteLine("Between dictionary and list : " + c2.Count);
	}
	
	class A
	{
		public string Something { get; set; }
	}
	
	class C
	{
		public string Name { get; set; }
		public string Value { get; set; }
	}
}
```

Cheers

**Comment by SteveWilkes** [comment]
Thanks for the report - I'll check it out. Mapping

```csharp
List<C> c2 = Mapper.Map(a.Values).ToANew<List<C>>();
```

...works, so it's something with the `ToTarget()` configuration.

Cheers,

Steve

**Comment by SteveWilkes** [comment]
This is fixed in the latest v1.7 release branch code - preview release to follow. I'm also going to add a `ToTargetInstead()` method to explicitly swap the source, which I assume is probably what you actually wanted to do!

Thanks again for the feedback!

Steve

**Comment by SteveWilkes** [comment]
This fix is included in v1.7-preview1, which is now available [on NuGet](https://www.nuget.org/packages/AgileObjects.AgileMapper).

Cheers!

**Comment by SteveWilkes** [comment]
This is all sorted in v1.7, which is now available [on NuGet](https://www.nuget.org/packages/AgileObjects.AgileMapper).

You can now use:

```csharp
Mapper
    .WhenMapping
    .FromDictionariesWithValueType<A>()
    .To<List<C>>()
    .Map((src, dst) => src.Values)
    .ToTargetInstead();
```
